### DISCUSSION TITLE
Support for much smaller MX1508 motor drivers?

### DISCUSSION THREAD

**Discussion by OMPEd** [mwood77/winderoo#53] [Ideas]
The L298N motor driver is quite large and I feel that I really have to shove it into the case to make it work, especially with the enclosed case remix.  It would be nice if support for PWM based motor driver boards like the MX1508 could also work.  The MX1508 boards are also less than half the size of the L298N.  I bought some MX1508 boards to test and found they work as a drop in replacement, but are being overdriven by the 5 or 6 volts from the ESP32 and spins WAY faster than normal because the digitalWrite only sets pin 25/26 to HIGH at system volatage.  If the code were updated to allow PWM control to fine-tune the motor control for smaller motor boards that would be awesome!

**Comment by mwood77** [comment]
I've thrown something together today - it builds OK, but I don't have an MX1508 on-hand to test. If you're willing to test it, I'd appreciate it.

To get the new code:
1. Pull the branch: `pwm-motor-control`
1. Then in your `platform.ini`, set the build flag `PWM_MOTOR_CONTROL` to `true`
1. Recompile the software and upload it to your winder

***

I've updated the documentation along with this, so take another look at [Flashing Your Microcontroller](https://github.com/mwood77/winderoo/blob/ac77a9b13c025dda027313dff3cb61ecb505e59c/docs/install-software.md#flashing-your-microcontroller), specifically step 3.
If you need to change the speed, go to [Troubleshooting](https://github.com/mwood77/winderoo/blob/ac77a9b13c025dda027313dff3cb61ecb505e59c/docs/install-software.md#troubleshooting).

**Comment by OMPEd** [comment]
Amazing, thanks!!  Will do some testing tonight and let you know!

**Comment by mwood77** [comment]
related: https://github.com/mwood77/winderoo/pull/54

**Comment by OMPEd** [comment]
Update: I tested the PWM branch and it worked great!  I tried PWM speeds from 100-255 and got a good feel for how the MX1508 drives the motor.  At a a value of 200 the motor rotates at ~23 RPM (not connected to the winder).  Comparing to a winder I built with the L298N which rotates at ~18 RPM at the motor, I think the notional value of the PWM setting in the code is around 145, but I'll assemble another winder and check compared to the original.

**Reply by mwood77** [reply]
Ah yes, that makes sense - the "belt reduction" is a 2:1 from what I recall.

**Reply by mwood77** [reply]
Come to think of it, I've actually made the rotation timing a user-adjustable variable in the code. 

From what I can remember (it's been a while...), this value is used to calculate the estimated winding time in `calculateWindingTime()`, which generates the estimated completion time, which is then passed to the Web UI.

You can adjust the value here: https://github.com/mwood77/winderoo/blob/ead282bc86873873eb266f47542487c576abd93e/src/platformio/osww-server/src/main.cpp#L38

**Reply by OMPEd** [reply]
Cool, I saw that variable last night but didn't play around with it.  I'm using my gear remix so my ratios are a little off from the standard OSWW, but I'll have to do some fine tuning in the code and the app to ensure the rotation is actually turning at 600TPD or whatever the UI says it should be.  With both the PWM setting and the durationInSecondsToCompleteOneRevolution variable I think I can fine tune it easily.  Maybe in a future release there could be a user-facing calibration section of the UI where the user puts in the time to complete 10 revolutions during a "calibration" event and the winder updates its parameters without having to reflash the ESP32.

**Reply by mwood77** [reply]
I'll add that kind of configuration stuff to my mental backlog - I've been wanting to do that; I'm very deep into the multi-winder meshing work right now, so it'll have to wait.

Let me know how tweaking `durationInSecondsToCompleteOneRevolution` works out.

**Comment by mwood77** [comment]
This feature has been merged into the main branch now.
